Feature branch merge

euqueme opened the pull request:
Requesting to merge on my development branch,

I disabled the cops: Metrics/CyclomaticComplexity, Metrics/MethodLength and Metrics/PerceivedComplexity
on my_all? my_any? my_none? and my_count? methods because they take arguments and are longer than 10 lines. 

Thank you in advance for your kind review.

euqueme committed:
MY_EACH, MY_EACH_WITH_INDEX AND MY_SELECT

Create feature_branch and delete feature_my_each and  feature_my_each_wtih_index branches to work all the features in this branch.

The code works with some made up test cases and has no rubocop offenses

euqueme committed:
MY_ALL

works with some dummy test cases but I had to disable some complexity metrics

euqueme committed:
MY_ANY

Passes some dummy test cases and also  had to disable complexity metrics

euqueme committed:
MY_NONE

Enumerable #6, passes all robocops

euqueme committed:
MY_COUNT

Passes some dummy test cases and it´s checked with rubocop so it has no offenses

euqueme committed:
MY_MAP

Passes some dummy test cases that had a range so added range validation to my_each and my_each_with_index methods

also passes all rubocops

euqueme committed:
MY_INJECT

Works with dummy cases and passes all rubocops

euqueme committed:
MULTIPLY_ELS

Method, passes all rubocops

euqueme committed:
MY_MAP WITH MY_PROC

modify it to take a proc but also takes a block if it is given.

euqueme committed:
PROC OVER BLOCK

Method my_map prefers to execute a Proc if both proc and block are given.

All cops checked

euqueme committed:
README

Update and add screenshots

euqueme committed:
COP FILES

and gitignore

euqueme committed:
DELETE METHODLENGTH

Metric because is not needed anymore

euqueme committed:
ENUMERABLES

is the name of the project

bolah2009 commented on main.rb:
```diff
@@ -0,0 +1,118 @@
+    size.times do |i|
+      is_a?(Range) ? yield(min + i) : yield(self[i])
+    end
+  end
```
`my_each` should return an enumerator when no block is given.

We could add a line that returns `to_enum` when no block is given.

> If no block is given, an enumerator is returned instead.

This applies to `my_each_with_index`, `my_select` and `my_map`.

bolah2009 commented on main.rb:
```diff
@@ -0,0 +1,118 @@
+      end
+    end
+    output
+  end
```
This line will result in an error when we pass an integer or string. For example: `[1,2,3].my_all?(2)` gives a `TypeError`

```ruby
          output = element.is_a?(arg) unless arg.is_a?(Regexp)
          output = element.match?(arg) if arg.is_a?(Regexp)
```
Let's handle cases when `arg` is not a class.
### Find more details about [`all?` here](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F)

bolah2009 commented on main.rb:
```diff
@@ -0,0 +1,118 @@
+      end
+    end
+    output
+  end
```
Please see the comment on `my_all?`

### Find more details about [`any?` here](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F)

bolah2009 commented on main.rb:
```diff
@@ -0,0 +1,118 @@
+      end
+    end
+    output
+  end
```
Please, see comment on `my_all?`

### Find more details about [`none?` here](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F)

[OPTIONAL]

I think if we could get the block from `my_none?` to `my_any?` we should only need something like `!my_any?` for `my_none?` to work fine. For example:

```ruby
  def my_none?(arg = nil, &block)
    !my_any?(arg = nil,&block)
  end
```

To make this work we have to use `block.call(element)` instead of `yield(element)` in `my_any?`

euqueme replied:
Solved! thank you very much for your reviews and suggestion I challenged myself and learned a lot.

Keep up the good work!!

bolah2009 replied:
Glad to hear 😀

bolah2009 commented on main.rb:
```diff
@@ -0,0 +1,118 @@
+    my_each_with_index { |ele, i| output = yield(output, ele) if i.positive? }
+    output = yield(output, arg) if arg
+    output
+  end
```
Let's handle cases where a `block` is not given and a `Symbol` or operator is passed instead. 

For example: `[1,2,3].my_inject(:+)` should equal `[1,2,3].inject(:+)`

HINT: Investigate the [`send`](https://ruby-doc.org/core-2.6.3/Object.html#method-i-send) method. Since `2.send(:+, 2) == 4`

bolah2009 reviewed (changes requested):
## Hey @euqueme   :wave: 

### This is an awesome effort :clap: :smile: 

Let's make changes to these methods (check comments for more details)
- `my_each?`
- `my_each_with_index?`
- `my_select?`
- `my_map?`
- `my_all?`
- `my_any?`
- `my_none?`
- `my_inject`

A reminder from pathwright:

> Take a look at Ruby docs about Enumerable class https://ruby-doc.org/core-2.6.5/Enumerable.html

#### :recycle:  Make these changes and submit another code review request. :+1: 

#### Feel free to reach me on slack if you having issues implementing these features. :+1: 

Good luck with your next submission. :star2: 

Cheers
[Bola](https://bolabuari.com/)

###### PS: Note that comments tagged as `OPTIONAL` are not part of the minimum requirements to approve this PR

euqueme committed:
ALL METHODS NOW

- my_each, my_each_with_index, my_map and my_select return an enumerable if no block is given
- my_any validates objects
- my_none is the negation of my_any
- my_inject takes symbols  or strings
- All cops pass

euqueme commented:
I fix all the changes requested by @bolah2009 

So right now:
- my_each, my_each_with_index, my_map and my_select return an enumerable if no block is given
- my_all and my_any validate objects 
- my_none is the negation of my_any
- my_inject takes symbols  or strings as parameters to operate the enumerable

Thank you in advance for your kind review.

samgaco commented on main.rb:
```diff
@@ -0,0 +1,131 @@
+        is_a?(Range) ? yield(min + i) : yield(self[i])
+      end
+    else
+      self
```
here `self` would be returning the array the method it's evaluating, for example:

```
irb(main):116:0> [1,2,3].my_each
=> [1, 2, 3]
```

but we want

```
irb(main):117:0> [1,2,3].each
=> #<Enumerator: [1, 2, 3]:each>
```

therefore, maybe we should transform self into a enumerator, something like `self.to_enum`  to get:

```
irb(main):118:0> [1,2,3].to_enum
=> #<Enumerator: [1, 2, 3]:each>
```

samgaco commented on main.rb:
```diff
@@ -0,0 +1,131 @@
+    elsif proc
+      my_each { |element| output << proc.call(element) }
+    else
+      return self
```
returning self again, it should be an enumerator what is returned for this method

```
irb(main):119:0> [1,2].map
=> #<Enumerator: [1, 2]:map>
```

samgaco commented on main.rb:
```diff
@@ -0,0 +1,131 @@
+      my_each { |element| yield(element) ? output.push(element) : nil }
+      output
+    else
+      self
```
should return an enumerator

samgaco commented on main.rb:
```diff
@@ -0,0 +1,131 @@
+        is_a?(Range) ? yield(min + i, i) : yield(self[i], i)
+      end
+    else
+      self
```
should be returning an enumerator

samgaco commented on main.rb:
```diff
@@ -0,0 +1,131 @@
+  def my_all?(arg = nil)
+    output = true
+    my_each do |element|
+      if output
```
you've got a gap in this method, check it out here compared as the original:

```
irb(main):123:0> ["abc",  "abc"].all?("abc")
=> true
irb(main):124:0> ["abc",  "abc"].my_all?("abc")
=> true
irb(main):125:0> ["abc",  1].all?("abc")
=> false
irb(main):126:0> ["abc",  1].my_all?("abc")
=> true
```

samgaco reviewed (changes requested):
Hi @euqueme :high_brightness: 

5 points to correct indicated in the comments below, please work on these changes and submit back for another code review once you are done.

Good spirit :muscle: 

---------------------

* ###### STATUS OF THE PROJECT: REQUESTED CHANGES :stop_sign:

euqueme committed:
ENUMERABLES RETURN

to_enum is return instead of self for my_each, my_each_with_index, my_select and my_map methods

Code passes all cops

euqueme committed:
MY_ALL

fix gap when elsif arg

euqueme commented:
I fix all the changes requested by @samgaco :

- All self returns were changed to to_enum 
- Fix gap in my_all for "elsif arg" so output can be set to either true or false

The code passes all cops.

Thank you in advance for your kind review.

AdnanAfsari reviewed (approved):
**Status: Approved**

Looks good now, its a 👍 from me.

Happy Coding
Adnan Afsari
